File: trees/binary_tree_paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        if node is None:
            return []
        elif node.left is None and node.right is None:
            return [[node.val]]
        else:
            allpaths = dfs(node.left) + dfs(node.right)
            for path in allpaths:
                path.insert(0, node.val)
            return allpaths

        return ['->'.join(map(str, path)) for path in dfs(root)]
        

File: trees/binary_tree_paths.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {String[]}
def binary_tree_paths(root)
    def dfs(node)
        if node.nil?
            []
        elsif node.left.nil? && node.right.nil?
            [[node.val]]
        else
            allpaths = dfs(node.left).concat(dfs(node.right))
            allpaths.each { |path| path.unshift(node.val)}
            allpaths
        end
    end

    dfs(root).map{ |path| path.join('->') }
end